connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTaskThread.java

```diff
@@ -50,16 +50,20 @@ public WorkerSinkTaskThread(WorkerSinkTask task, String name, Time time,
 
     @Override
     public void execute() {
-        // Try to join and start. If we're interrupted before this completes, bail.
-        if (!task.joinConsumerGroupAndStart())
-            return;
+        try {
+            // Try to join and start. If we're interrupted before this completes, bail.
+            if (!task.joinConsumerGroupAndStart())
+                return;
 
-        while (getRunning()) {
-            iteration();
-        }
+            while (getRunning()) {
+                iteration();
+            }
 
-        // Make sure any uncommitted data has committed
-        task.commitOffsets(true, -1);
+            // Make sure any uncommitted data has committed
+            task.commitOffsets(true, -1);
+        } catch (Throwable t) {
+            task.abort(t);
```

Comment:
Since the Worker currently doesn't get any notification when a task fails, it seemed necessary to cleanup the task state, but I agree that it's quite confusing. One thing that makes this code a bit difficult to work with is the unclear boundary between WorkerSinkTask and WorkerSinkTaskThread. The methods overridden from WorkerTask are all intended to be called from the Worker's thread, but the rest are reserved only for the task thread. It's especially difficult to tell what state is accessed only from the task thread (which is probably the reason for the seemingly unnecessary synchronization). 

This would be a larger effort (and well beyond the scope of this ticket), but I think what I'd like to do is something like the following:

1. Encapsulate everything that the task thread needs in WorkerSinkTaskThread. This means moving the better part of WorkerSinkTask into WorkerSinkTaskThread. The only thing left in WorkerSinkTask would be the code to start and stop the thread. 
2. Generalize what is left of WorkerSinkTask into an abstract class which can also be used for WorkerSourceTask. Then we could basically do WorkerSinkTask -> AbstractWorkerTask, and WorkerSinkTaskThread -> WorkerSinkTask. (In fact, the easier transformation is probably to move the methods from WorkerSinkTaskThread into WorkerSinkTask and then pull up the overridden methods of WorkerTask into AbstractWorkerTask.)
3. Have WorkerSinkTaskThread implement Runnable instead of extending Thread. This would be required to have WorkerSinkTask extend AbstractWorkerTask. Runnables tend to be easier to test, and it opens the door for using an executor if we ever have a reason to. The only thing we'd lose is the ShutdownableThread, but it wouldn't be difficult to adapt to still make use of this class.

If you did all of this, what you'd be left with is a WorkerSinkTask which contains all its state and is at least as easy to test as it was before. You'd also be left with a little less common boilerplate in WorkerSinkTask and WorkerSourceTask and we'd no longer need WorkerSinkTaskThread and WorkerSourceTaskThread.

But for this ticket, maybe I can just remove abort() and we can leave it to the Worker to eventually cleanup task state (as in the current implementation)?